Cleaning up the GoSource provider usage

ianthehat opened the pull request:
We use it to manage all source lists, and propagate both original and modified
sources sets from rules.
This is a preparatory step for cleaning up both the library attribute and the
cover rules.

jayconrod commented on go/private/common.bzl:
```diff
@@ -68,3 +68,29 @@ def pkg_dir(workspace_root, package_name):
     return package_name
   return "."
+
+def dict_of(st):
```
Can this be private?

jayconrod commented on go/private/common.bzl:
```diff
@@ -68,3 +68,29 @@ def pkg_dir(workspace_root, package_name):
+      data[key] = value
+  return data
+
+def merge(a, b):
```
It seems strange for `merge` and `replace` to accept structs and return dicts. What do you think about taking a constructor as another argument and returning a struct?

jayconrod commented on go/private/common.bzl:
```diff
@@ -68,3 +68,29 @@ def pkg_dir(workspace_root, package_name):
+def replace(a, b):
+  """Returns a dict where values from b are used when present in both a and b"""
+  data = dict_of(a)
+  for key, value in dict_of(b).items():
```
I think this is equivalent to `data.update(dict_of(b))`

ianthehat commented:
I am dropping this for #699.
Some of it is the same, but I removed the GoSource provider instead of improving it, pass around flat lists instead of exploded forms, and dropped replace and merge completely.
